From 4f4e6609bb4992f8694a9063db6a27e485ac0627 Mon Sep 17 00:00:00 2001
From: Jason Gerecke <killertofu@gmail.com>
Date: Mon, 4 Aug 2014 15:12:40 -0700
Subject: [PATCH] ISDv4 touch query timeout should be non-fatal

Both the wacom_w8001 and isdv4.c drivers assume that a tablet /may/
respond to the touchquery packet with silence. This assumption does
not carry over to tools-shared.c, and may cause pen-only tablet PC
users trouble with isdv4-serial-inputattach.

Signed-off-by: Jason Gerecke <killertofu@gmail.com>
---
 tools/tools-shared.c | 42 +++++++++++++++++++++++-------------------
 1 file changed, 23 insertions(+), 19 deletions(-)

diff --git a/tools/tools-shared.c b/tools/tools-shared.c
index 10d7008..cdd7621 100644
--- a/tools/tools-shared.c
+++ b/tools/tools-shared.c
@@ -303,27 +303,31 @@ int query_tablet(int fd)
 	TRACE("Trying touch query\n");
 	if (stop_tablet(fd)) goto out;
 	if (write_to_tablet(fd, ISDV4_TOUCH_QUERY)) goto out;
-	if (wait_for_tablet(fd)) goto out;
-
-	memset(buffer, 0, sizeof(buffer));
-	len = read_data(fd, buffer, ISDV4_PKGLEN_TPCCTL);
-	if (len < 1 && errno != EAGAIN)
-		goto out;
-
-	TRACE("Parsing touch query reply.\n");
-	rc = isdv4ParseTouchQuery(buffer, len, &touch);
-	if (rc < 0)
-	{
-		fprintf(stderr, "touch parsing error code %d\n", rc);
-		/* failure to parse touch query is not fatal */
-	} else {
-		printf("TOUCH: version: %d\n", touch.version);
-		printf("TOUCH: x max: %d y max %d\n", touch.x_max, touch.y_max);
-		printf("TOUCH: panel resolution: %d\n", touch.panel_resolution);
-		printf("TOUCH: capacity resolution: %d\n", touch.capacity_resolution);
-		printf("TOUCH: sensor id: %d\n", touch.sensor_id);
+	if (wait_for_tablet(fd) < 0) {
+		fprintf(stderr, "ignoring touch query timeout\n");
+		touch.sensor_id = 0;
+		/* failure to recieve reply to touch query is not fatal */
 	}
+	else {
+		memset(buffer, 0, sizeof(buffer));
+		len = read_data(fd, buffer, ISDV4_PKGLEN_TPCCTL);
+		if (len < 1 && errno != EAGAIN)
+			goto out;
 
+		TRACE("Parsing touch query reply.\n");
+		rc = isdv4ParseTouchQuery(buffer, len, &touch);
+		if (rc < 0)
+		{
+			fprintf(stderr, "touch parsing error code %d\n", rc);
+			/* failure to parse touch query is not fatal */
+		} else {
+			printf("TOUCH: version: %d\n", touch.version);
+			printf("TOUCH: x max: %d y max %d\n", touch.x_max, touch.y_max);
+			printf("TOUCH: panel resolution: %d\n", touch.panel_resolution);
+			printf("TOUCH: capacity resolution: %d\n", touch.capacity_resolution);
+			printf("TOUCH: sensor id: %d\n", touch.sensor_id);
+		}
+	}
 	return touch.sensor_id;
 
 out:
-- 
2.0.4

